generator/src/FluentStageFactoryGenerator.php

```diff
@@ -0,0 +1,163 @@
+    {
+        $namespace = new PhpNamespace($definition->namespace);
+        $class = $namespace->addClass('FluentFactory');
+
+        $namespace->addUse(StageInterface::class);
+        $namespace->addUse(FieldQueryInterface::class);
+        $namespace->addUse(Pipeline::class);
+        $namespace->addUse(Decimal128::class);
+        $namespace->addUse(Document::class);
+        $namespace->addUse(Int64::class);
+        $namespace->addUse(PackedArray::class);
+        $namespace->addUse(Serializable::class);
+        $namespace->addUse(Timestamp::class);
+        $namespace->addUse(Type::class);
+        $namespace->addUse(ArrayFieldPath::class);
+        $namespace->addUse(FieldPath::class);
+        $namespace->addUse(ResolvesToArray::class);
+        $namespace->addUse(ResolvesToObject::class);
+        $namespace->addUse(Pipeline::class);
+        $namespace->addUse(AccumulatorInterface::class);
+        $namespace->addUse(ExpressionInterface::class);
+        $namespace->addUse(Optional::class);
+        $namespace->addUse(QueryInterface::class);
+        $namespace->addUse(Sort::class);
+        $namespace->addUse(BSONArray::class);
+        $namespace->addUse(stdClass::class);
+        $namespace->addUse(self::FACTORY_CLASS);
+        $class->addProperty('pipeline')
+            ->setType('array')
+            ->setComment('@var list<StageInterface>')
+            ->setValue([]);
+        $class->addMethod('getPipeline')
+            ->setReturnType(Pipeline::class)
+            ->setBody(<<<'PHP'
+                return new Pipeline(...$this->pipeline);
+                PHP);
+
+        $staticFactory = ClassType::from(self::FACTORY_CLASS);
+        assert($staticFactory instanceof ClassType);
+        foreach ($staticFactory->getMethods() as $method) {
+            if (! $method->isPublic()) {
+                continue;
+            }
+
+            try {
+                $this->addMethod($method, $namespace, $class);
+            } catch (Throwable $e) {
+                throw new RuntimeException(sprintf('Failed to generate class for operator "%s"', $operator->name), 0, $e);
+            }
+        }
+
+        $staticFactory = TraitType::from(Stage\FactoryTrait::class);
+        assert($staticFactory instanceof TraitType);
+        foreach ($staticFactory->getMethods() as $method) {
+            if (! $method->isPublic()) {
+                continue;
+            }
+
+            try {
+                $this->addMethod($method, $namespace, $class);
+            } catch (Throwable $e) {
+                throw new RuntimeException(sprintf('Failed to generate class for operator "%s"', $operator->name), 0, $e);
+            }
+        }
+
+        return $namespace;
+    }
+
+    private function addMethod(Method $factoryMethod, PhpNamespace $namespace, ClassType $class): void
+    {
+        if ($class->hasMethod($factoryMethod->getName())) {
+            return;
```

Comment:
Would it make sense to throw here? What is the legitimate case where the method already exists and you'd want to NOP?

Reply:
The method `match` is redefined in the `Stage` class. The generated version from the trait must be ignored.

Reply:
Noted. In this case, a comment would be helpful to clarify that you're giving preference to the redefined Stage methods instead of those from the trait. And this assumes you add the Stage methods _before_ the trait (no need to repeat that, it's clear from the order of `addMethod()` calls in the generation method).

Reply:
That's essentially what I said in the comment above.

Reply:
Oh, my mistake. I just realized this entire thread was incorrectly applied to the early return for non-public methods.

I see the comment below for skipping overridden methods. :+1: